File: src/main/scala/com/teambrmodding/assistedprogression/managers/RecipeManager.scala
package com.teambrmodding.assistedprogression.managers

import java.util

import com.teambrmodding.assistedprogression.registries.GrinderRecipeHandler
import net.minecraft.command.ServerCommandManager

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis <pauljoda>
  * @since 2/20/2016
  */
object RecipeManager {
    sealed trait RecipeType { def name : String }
    case object Grinder     extends RecipeType { val name = "grinder" }

    lazy val recipeHandlers = new util.HashMap[RecipeType, AbstractRecipeHandler[_, _, _]]()

    def preInit() : Unit = {
        CraftingRecipeManager.preInit()
    }

    def init() : Unit = {
        recipeHandlers.put(Grinder,    new GrinderRecipeHandler().loadHandler())
    }

    def initCommands(manager : ServerCommandManager) : Unit = {
        for(handlerType <- recipeHandlers.keySet().toArray) {
            val command = recipeHandlers.get(handlerType.asInstanceOf[RecipeType]).getCommand
            if(command != null)
                manager.registerCommand(command)
        }
    }

    /**
      * Used to get a handler Type, this way we know only defined handlers are cast and sent back, ensures
      * no bad casting (if done correctly)
      *
      * @param handlerType The handler Type
      * @tparam H The Handler Object
      * @return The Handler
      */
    def getHandler[H <: AbstractRecipeHandler[_, _, _]](handlerType : RecipeType) : H =
        recipeHandlers.get(handlerType).asInstanceOf[H]

    /**
      * Used to get Handles by name
      *
      * @param name The name of the handler, eg "Crusher"
      * @tparam H The recipe handler type
      * @return The recipe Handler
      */
    def getHandler[H <: AbstractRecipeHandler[_, _, _]](name : String)  : Option[H] = {
        for(x <- recipeHandlers.keySet().toArray) {
            val handler = x.asInstanceOf[RecipeType]
            if(handler.name.equalsIgnoreCase(name))
                return Option(recipeHandlers.get(handler).asInstanceOf[H])
        }
        None
    }
}

File: src/main/java/com/teambrmodding/assistedprogression/managers/RecipeManager.java
package com.teambrmodding.assistedprogression.managers;

import com.teambrmodding.assistedprogression.registries.AbstractRecipeHandler;
import com.teambrmodding.assistedprogression.registries.GrinderRecipeHandler;
import gnu.trove.map.hash.THashMap;
import net.minecraft.command.CommandBase;
import net.minecraft.command.ServerCommandManager;

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis - pauljoda
  * @since 2/15/2017
  */
public class RecipeManager {

    // The list of all recipe handlers
    public static THashMap<RecipeType, AbstractRecipeHandler> recipeHandlers = new THashMap<>();

    /**
      * Called to register things
      */
    public static void preInit() {
        CraftingRecipeManager.preInit();
    }

    /**
      * Builds the recipe registry
      */
    public static void init() {
        recipeHandlers.put(RecipeType.GRINDER, new GrinderRecipeHandler().loadHandler());
    }

    /**
      * Loads all the commands from the handlers
      * @param manager The manager to register to
      */
    public static void initCommands(ServerCommandManager manager) {
        for(RecipeType recipeType : recipeHandlers.keySet()) {
            CommandBase command = recipeHandlers.get(recipeType).getCommand();
            if(command != null)
                manager.registerCommand(command);
        }
    }

    /*******************************************************************************************************************
      * Helper Methods                                                                                                  *
      *******************************************************************************************************************/

    /**
      * Used to get a recipe handler
      * @param recipeType The recipe type
      * @return The recipe handler
      */
    public static  <H extends AbstractRecipeHandler> H getHandler(RecipeType recipeType) {
        return (H) recipeHandlers.get(recipeType);
    }

    public enum RecipeType {
        GRINDER("grinder");

        private String name;

        RecipeType(String name) {
            this.name = name;
        }

        /**
          * Get the name of this recipe type
          * @return The name
          */
        public String getName() {
            return name;
        }
    }
}